File: src/jvmception/objects/cp/CpString.java
package jvmception.objects.cp;

import java.io.DataInputStream;
import java.io.IOException;
import java.lang.ref.Reference;

import jvmception.jvmtypes.IUnitSerializable;
import jvmception.jvmtypes.JVMReference;
import jvmception.objects.JVMInstance;

public class CpString extends CpInfo{
	private int stringIndex;
	private JVMReference reference = null;
	private CpInfo[] cpInfo;
	
	public CpString(DataInputStream dis, CpInfo[] cpInfo) throws IOException{
		stringIndex = dis.readUnsignedShort();
		this.cpInfo = cpInfo;
	}
	
	public int getStringIndex() {
		return stringIndex;
	}

	@Override
	public IUnitSerializable toIUnitSerializable() {
		if (reference == null) {
			cpInfo = null;
			reference = (JVMReference) ((CpUtf8)cpInfo[stringIndex]).toIUnitSerializable();
		}
		return reference;
	}
}

File: src/jvmception/objects/cp/CpString.cs


using System.IO;

public class CpString : CpInfo
{
    private int stringIndex;
    private JVMReference reference = null;
    private CpInfo[] cpInfo;

    public CpString(BinaryReaderBigEndian dis, CpInfo[] cpInfo)
    {
        stringIndex = dis.ReadUInt16();
        this.cpInfo = cpInfo;
    }

    public int getStringIndex()
    {
        return stringIndex;
    }

    public override IUnitSerializable toIUnitSerializable()
    {
        if (reference == null)
        {
            cpInfo = null;
            reference = (JVMReference)((CpUtf8)cpInfo[stringIndex]).toIUnitSerializable();
        }
        return reference;
    }
}
